src/main/java/it/smartcommunitylab/aac/tos/TosOnAccessFilter.java

```diff
@@ -0,0 +1,160 @@
+
+	private final RealmManager realmManager;
+	private final MyUserManager userManager;
+
+	public TosOnAccessFilter(RealmManager realmManager, MyUserManager userManager) {
+		// init request cache as store
+		HttpSessionRequestCache cache = new HttpSessionRequestCache();
+		this.requestCache = cache;
+		this.realmManager = realmManager;
+		this.userManager = userManager;
+		this.requestMatcher = buildRequestMatcher();
+	}
+
+	private RequestMatcher buildRequestMatcher() {
+		List<RequestMatcher> antMatchers = Arrays.stream(SKIP_URLS).map(u -> new AntPathRequestMatcher(u))
+				.collect(Collectors.toList());
+
+		return new NegatedRequestMatcher(new OrRequestMatcher(antMatchers));
+
+	}
+
+	public void setRedirectStrategy(RedirectStrategy redirectStrategy) {
+		this.redirectStrategy = redirectStrategy;
+	}
+
+	public void setRequestMatcher(RequestMatcher requestMatcher) {
+		this.requestMatcher = requestMatcher;
+	}
+
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+			throws ServletException, IOException {
+		if (requestMatcher.matches(request) && requiresProcessing(request)
+				&& !termsManagedRequestMatcher.matches(request)) {
+			logger.trace("process request for {}", request.getRequestURI());
+			UserAuthentication userAuth = (UserAuthentication) SecurityContextHolder.getContext().getAuthentication();
+			ExtendedAuthenticationToken token = CollectionUtils.firstElement(userAuth.getAuthentications());
+
+			if (token == null) {
+				logger.error("empty token on authentication success");
+				return;
+			}
+
+			String realm = token.getRealm();
+
+			try {
+				User user = userManager.getMyUser(realm);
+				if (!realm.equalsIgnoreCase(SystemKeys.REALM_GLOBAL) && !realm.equalsIgnoreCase(SystemKeys.REALM_SYSTEM)
+						&& !user.isTosAccepted()) {
+					Realm realmEntity = realmManager.findRealm(realm);
+
+					if ((realmEntity.getTosConfiguration().getConfiguration().containsKey("enableTOS"))
+							&& (boolean) realmEntity.getTosConfiguration().getConfiguration().get("enableTOS")
+							&& request.getSession().getAttribute("termsManaged") == null) {
+						String targetUrl = "/terms/"
+								+ realmEntity.getTosConfiguration().getConfiguration().get("approveTOS");
+						this.requestCache.saveRequest(request, response);
+						this.logger.debug("Redirect to {}", targetUrl);
+						this.redirectStrategy.sendRedirect(request, response, targetUrl);
+						return;
+					} else if (request.getSession().getAttribute("termsManaged") != null
+							&& request.getSession().getAttribute("termsManaged").equals("true")) {
+						SavedRequest savedRequest = this.requestCache.getRequest(request, response);
+						if (savedRequest != null) {
+							logger.debug("restore request from cache");
+							this.requestCache.removeRequest(request, response);
+							this.redirectStrategy.sendRedirect(request, response, savedRequest.getRedirectUrl());
```

Comment:
before performing the redirect you also need to check if the user actually *accepted* the terms, otherwise you don't restore the request but bounce him to the error page. Also you should clear the security context if rejected and the terms are *required*